projects/demo/src/app/forms/dynamic-controls/dynamic-controls.demo.ts

```diff
@@ -4,128 +4,39 @@
-  private registerPlatformChange() {
-    this.form.controls.platform.valueChanges.subscribe(value => {
-      this.form.removeControl('s3');
-      this.form.removeControl('iam');
-      this.form.removeControl('console');
-
-      switch (value) {
-        case 'CLOUDIAN':
-          this.addControlsForCloudian();
-          break;
-        case 'ECS':
-          this.addControlsForECS();
-          break;
-        default:
-          break;
-      }
-
-      this.setCertValidationValue();
-    });
-  }
-
-  private addControlsForCloudian() {
-    this.addS3Control('cloudian signature');
-    this.addIAMControl();
-  }
-
-  private addControlsForECS() {
-    this.addS3Control();
-    this.addConsoleControl();
-  }
-
-  private addS3Control(signature?: string) {
-    this.form.addControl(
-      's3',
-      this.fb.group({
-        url: this.fb.control('', [Validators.required]),
-        tlsSignature: this.fb.control(
-          {
-            value: signature,
-            disabled: true,
-          },
-          Validators.required
-        ),
-      })
-    );
-  }
-
-  private addIAMControl() {
-    this.form.addControl(
-      'iam',
-      this.fb.group({
-        url: this.fb.control('', [Validators.required]),
-        tlsSignature: this.fb.control(
-          {
-            value: '',
-            disabled: true,
-          },
-          Validators.required
-        ),
-      })
-    );
-  }
-
-  private addConsoleControl() {
-    this.form.addControl(
-      'console',
-      this.fb.group({
-        url: ['', [Validators.required]],
-        secretKey: ['', [Validators.required]],
-      })
-    );
-  }
-
-  // set radio based on form value
-  private setCertValidationValue() {
-    const signatureControl = this.form.controls['s3'].get('tlsSignature') as FormControl;
-
-    const value = signatureControl.value ? 'tlsSignature' : 'publicSigned';
+  constructor(private cd: ChangeDetectorRef) {}
 
-    this.certValidation.setValue(value);
-  }
-
-  private registerCertValidateChange() {
-    this.certValidation.valueChanges.subscribe(value => {
-      const tlsSignature = this.form.controls['s3'].get('tlsSignature') as FormControl;
-
-      if (value === 'tlsSignature') {
-        tlsSignature.enable({ emitEvent: false });
-      } else {
-        // public signed
-        tlsSignature.disable({ emitEvent: false });
-      }
-
-      setTimeout(() => {
-        if (tlsSignature.value) {
-          tlsSignature.markAsTouched();
-          tlsSignature.updateValueAndValidity();
+  ngOnInit() {
+    this.form
+      .get('platform')
+      ?.valueChanges.pipe(takeUntil(this.destroyed))
+      .subscribe(res => {
+        console.log(res);
+        if (res === 'two') {
+          this.form.addControl('control', new FormControl('test'));
+          this.form.get('control')?.markAsTouched();
+          this.form.get('control')?.updateValueAndValidity();
         }
       });
-    });
+  }
+
+  ngOnDestroy() {
+    this.destroyed.complete();
```

Comment:
```suggestion
    this.destroyed.next();
    this.destroyed.complete();
```

The `takeUntil` operator will only unsubscribe when the `notifier` emits a value.